stream.go

```diff
@@ -132,6 +136,43 @@ func (s *Stream) ReadSCTP(p []byte) (int, PayloadProtocolIdentifier, error) {
 	}
 }
 
+// SetReadDeadline sets the read deadline in an identical way to net.Conn
+func (s *Stream) SetReadDeadline(deadline time.Time) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.readTimeoutCancel != nil {
+		close(s.readTimeoutCancel)
+		s.readTimeoutCancel = nil
+	}
+
+	if deadline.IsZero() {
+		if errors.Is(s.readErr, errReadDeadlineExceeded) {
+			s.readErr = nil
+		}
+	} else {
+		s.readTimeoutCancel = make(chan struct{})
+
+		go func(readTimeoutCancel chan struct{}) {
+			t := time.NewTimer(time.Until(deadline))
+			select {
+			case <-readTimeoutCancel:
```

Comment:
I think this channel should be closed when s.readErr is set to a non-nil value outside this function so that this goroutine would exit right away. Any thoughts?

Reply:
Agreed, updated. Thanks!

Reply:
Thanks @nerd2!

I have just noticed there is [one other path](https://github.com/nerd2/sctp/blob/396e5d548cb3fdecec4f2dc02e19b60c5c842286/association.go#L606) that sets s.readError to a non-nil value.

I'm wondering if we could close the channel in one place using defer inside s.ReadSCTP method...

```go
func (s *Stream) ReadSCTP(p []byte) (int, PayloadProtocolIdentifier, error) {
	s.lock.Lock()
	defer s.lock.Unlock()

        defer func() {
                // close readTimeoutCancel if not nil while the lock is still held
                if s.readTimeoutCancel != nil {
                        close(s.readTimeoutCancel)
                        s.readTimeoutCancel = nil
                }
        }()

	for {
```

Reply:
I believe even if a single (or multiple) ReadSCTP succeeds, the definition of SetReadDeadline requires the read timeout routine still to run so it applies to following ReadSCTPs.

BTW I did update association.go too

Reply:
Yes, I agree. As I forgot to add `&& s.readErr != nil` to the if condition in my code snippet, let me ask your thought once again.

```go
func (s *Stream) ReadSCTP(p []byte) (int, PayloadProtocolIdentifier, error) {
	s.lock.Lock()
	defer s.lock.Unlock()

        defer func() {
                // close readTimeoutCancel if the current read timeout routine is no longer effective
                if s.readTimeoutCancel != nil && s.readErr != nil {
                        close(s.readTimeoutCancel)
                        s.readTimeoutCancel = nil
                }
        }()

	for {
```

Note:
* If s.readErr != nil, subsequent ReadSCTPs would return immediately with the error

Reply:
Added, and removed readTimeoutCancel handling from other readErr cases